src/main/java/ui/components/pickers/LabelPickerUILogic.java

```diff
@@ -8,6 +8,28 @@
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
+/**
+ * Handles the logic of the label picker UI.
+ *
+ * The crux of the logic is as follows:
+ * 1. When the text in the LabelPicker's text field changes, processTextFieldChange is called.
+ * 2. If the text is not empty, this method will try to determine if the the user typed in a character or backspaced.
+ *   2a. Typing is handled by addCharHandler
+ *     i. If the user typed ' ', it will attempt to add the current highlighted label (if there is).
+ *     ii. Else it is an ongoing query
+ *     Bottom labels that do not fit the query will be faded.
+ *     The first of the unfaded labels will be 'highlighted'.
+ *     The highlighted bottom label will then be displayed as a 'possible label' at the top.
+ *   2b. Backspacing is handled by removeCharHandler
+ *     i. When the current last character is ' ', there is no current query, therefore any 'possible label'
+ *     is removed
+ *     ii. When the removed character is ' ', and the last character is not ' ', the most recently specified label is
+ *     toggled. In addition, there is an existing query that will result in a 'possible label'
+ *
+ * listOfSpecifiedLabels is maintained to keep track of the mapping between what the user has entered in the
+ * text field, and the labels that have been added/removed due to them.
+ * listOfRemovedExclusiveLabels is also maintained so that the removed labels can be added back when needed.
+ */
```

Comment:
Need for extensive comments to explain the algorithm is an indication that the code is not self-explanatory. Consider this code:
```java
if(text.isEmpty()) return;

if(isBackSpace(text){
    handleBackspace(text);
} else if (isSpace(text)){
    handleSpace(text);
} else {
    handleChar(text);
}
```
Now the high level logic is clear and there is no need for comments. Use the same approach for each of the methods called. In other words, apply SLAP. The comment should explain WHAT the method does and WHY it does it, but not HOW it does it (the HOW should be clear from the code itself).